lib/ui/common/RegExInputFormatter.dart

```diff
@@ -0,0 +1,45 @@
+import 'package:flutter/services.dart';
+
+class RegExInputFormatter implements TextInputFormatter {
+  final RegExp _regExp;
+
+  RegExInputFormatter._(this._regExp);
+
+  factory RegExInputFormatter.withRegex(String regexString) {
+    try {
+      final regex = RegExp(regexString);
+      return RegExInputFormatter._(regex);
+    } catch (e) {
+      // Something not right with regex string.
+      assert(false, e.toString());
+      return null;
+    }
+  }
+
+  @override
+  TextEditingValue formatEditUpdate(
+      TextEditingValue oldValue, TextEditingValue newValue) {
+    final oldValueValid = _isValid(oldValue.text);
+    final newValueValid = _isValid(newValue.text);
+    if (oldValueValid && !newValueValid) {
+      return oldValue;
+    }
+    return newValue;
+  }
+
+  bool _isValid(String value) {
+    try {
+      final matches = _regExp.allMatches(value);
+      for (Match match in matches) {
```

Comment:
could we just use hasMatch instead of the loop, or (stringMatch == the original)? feels like this could be simpler? what's the reason for the loop?